cpp/vm/evmzero/memory_test.cc

```diff
@@ -214,5 +214,73 @@ TEST(MemoryTest, Equality) {
   EXPECT_EQ(m1, m2);
 }
 
+struct TestParams {
+  const size_t size;
+  const size_t start_offset;
+  const ptrdiff_t dest_offset;
+};
+class ParametrizedMemoryTest : public testing::TestWithParam<TestParams> {};
+
+constexpr auto kMB = 1024 * 1024;
+constexpr auto kBufferSize = 16 * kMB;
+
+INSTANTIATE_TEST_SUITE_P(ParametrizedMemoryTestWithSizes, ParametrizedMemoryTest,
+                         testing::Values(
+                             TestParams{
+                                 .size = 4 * kMB,
+                                 .start_offset = 0,
+                                 .dest_offset = 0,
+                             },
+                             TestParams{
+                                 .size = 4 * kMB,
+                                 .start_offset = 2 * kMB,
+                                 .dest_offset = 0,
+                             },
+                             TestParams{
+                                 .size = 4 * kMB,
+                                 .start_offset = 0,
+                                 .dest_offset = 2 * kMB,
+                             },
+                             TestParams{
+                                 .size = 4 * kMB,
+                                 .start_offset = kBufferSize - 1,
+                                 .dest_offset = 0,
+                             },
+                             TestParams{
+                                 .size = 4 * kMB,
+                                 .start_offset = 0,
+                                 .dest_offset = kBufferSize - 1,
+                             }));
+
+TEST_P(ParametrizedMemoryTest, MemCopy) {
+  const auto [size, start_offset, dest_offset] = GetParam();
+  const auto effective_size = std::min(kBufferSize - start_offset, size);
+
+  std::vector<uint8_t> data(kBufferSize);
+  std::fill(data.begin(), data.end(), 0);
+
+  // Fill memory to be copied with a pattern not congruence of 2,4,8,16... to check for aliasing
+  auto count = 0ul;
+  auto to_fill = std::span{data.data() + start_offset, effective_size};
+  std::transform(to_fill.begin(), to_fill.end(), to_fill.begin(), [&](uint8_t) { return count++ % 37; });
+
+  Memory memory(data);
+  EXPECT_EQ(memory.GetSize(), kBufferSize);
+  auto check_pattern = [&](std::span<uint8_t> span) {
+    int count = 0ul;
+    for (auto i = 0ul; i < effective_size; ++i) {
+      EXPECT_EQ(span[i], count++ % 37) << "at index " << i;
+    }
+  };
+
+  // Check pattern before copy in source range
+  check_pattern(memory.GetSpan(start_offset, size));
+
+  memory.MemCopy(0, start_offset, size);
```

Comment:
This will potentially grow the memory, which should be checked.
Also, if so, the `effective_size` will be to short (just 1 byte long for some tests) and the `check_pattern` will not check all the copied data, right?

Reply:
Right, the remainder of the copied data should be zeroes (assuming zero initialized memory expansions) 
if we did an out of bounds copy, I will check for zeroes.